### ISSUE TITLE
[GLES2] Repeated textures look blurry in HTML5 builds

### ISSUE THREAD

**Issue by theogen-ratkin** [godotengine/godot#44379]
**Godot version:**
`3.2.3.stable.official`

**OS/device including version:**

The issue happens on HTML5 with GLES2 backend. Tested on Arch Linux with Firefox 83 and Chromium 87.0.4280.88.

Tested on two different computers, one has Intel i5-3320M and the other has NVIDIA GTX 1650 SUPER.

**Issue description:**

If you set "Repeat" to enabled in texture import settings, the texture will look blurry in HTML5 build, like it has been scaled with aliasing. This causes semitransparent edges and generally blurry look for pixel art.

The issue happens with GLES2 backend, it's not present on GLES3.

How it's supposed to look:
![1](https://user-images.githubusercontent.com/20743993/102134649-39da7280-3e2d-11eb-9fd8-7f7dd2b3de09.png)

How it looks instead:
![2](https://user-images.githubusercontent.com/20743993/102134655-3cd56300-3e2d-11eb-9146-f7c570682f17.png)

**Steps to reproduce:**

1. Add any pixel art texture to you project and make sure that you can see it up close.
2. Disable filtering and enable repeat in import settings and click "Reimport".
![image](https://user-images.githubusercontent.com/20743993/102137262-ebc76e00-3e30-11eb-93d9-7b9c2e616876.png)
3. Export the project using HTML5 preset.
4. Run it in your browser. You will likely need to start a local HTTP server like this: `python -m http.server --bind 127.0.0.1`

**Minimal reproduction project:** [blurred_repeated_pixels.zip](https://github.com/godotengine/godot/files/5691552/blurred_repeated_pixels.zip)

There's already an HTML5 build in the `build/` directory.

**Comment by Calinou** [comment]
I can confirm this on Firefox 84 and Chromium 86 (Fedora 33, GeForce GTX 1080).

**Comment by clayjohn** [comment]
Can either of you try with high precision floats enabled?

https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-rendering-gles2-compatibility-enable-high-float-android

**Comment by theogen-ratkin** [comment]
@clayjohn Just tried it, it doesn't fix the issue.

**Comment by lawnjelly** [comment]
Always try this kind of thing with and without batching BTW as that can help pin it down. It is probably due to wrapping with a non power of two texture. Top tip, use POT textures on HTML5 / mobile, especially when using any kind of wrapping.

I'll see if I can reproduce. :+1:

**Comment by theogen-ratkin** [comment]
@lawnjelly Can confirm that using a power of two texture fixes the problem! Ironically, all textures in my project were power of two except for repeated ones. :man_facepalming:

**Comment by theogen-ratkin** [comment]
@larsonjj Is it a driver issue or can it be fixed within the engine? GLES3 doesn't have this problem.

**Comment by lawnjelly** [comment]
GLES3 mandates non POT texture support, GLES2 does not. Hence hacky workarounds are used on GLES2 which may look bad and run like a dog. But you can avoid this problem by just using POT textures.

**Comment by clayjohn** [comment]
I wonder if our POT codepath accidentally forces texture filtering. 

@theogen-ratkin can you compare the texture displaying the issue with filter on and off?

**Comment by theogen-ratkin** [comment]
@lawnjelly Oh I see. I agree that it's better to just adhere to the standard, but I think it would be nice if Godot had some fallback option. Maybe resize textures at export time?

@clayjohn No, with filtering the texture looks completely blurry. But you can see that the situation is the same: original image is consistent in color, the second one has semitransparent edges.

![image](https://user-images.githubusercontent.com/20743993/102245672-38b05080-3ed4-11eb-9ec7-27f0bc3a87d0.png)

![image](https://user-images.githubusercontent.com/20743993/102245736-4a91f380-3ed4-11eb-94dc-6a0b814cc7d5.png)

**Comment by theogen-ratkin** [comment]
> What you are seeing _is_ the fallback option

Fair enough! I just think that it can really hurt Godot's adoption, since people coming from, say, Unity, expect things to "just work" the same on all platforms, no matter what you throw at it. And even though there's Vulkan, GLES2 is not going away anytime soon.

> Hence hacky workarounds are used on GLES2 which may look bad and run like a dog.

What kind of workarounds? Are they really so bad? If so, we could make those workarounds optional, so that developers could decide for themselves if they want them.

**Comment by clayjohn** [comment]
We do resize the texture at load time https://github.com/godotengine/godot/blob/6956b01cb55a082440df044211c2e86427b6c62c/drivers/gles2/rasterizer_storage_gles2.cpp#L677-L686

I'm guessing that the blurriness is actually caused by the resize rather than the custom repeat shader code.

**Comment by theogen-ratkin** [comment]
@clayjohn If that's the case then it's great! I thought that it looked like it just has been resized with filtering.

> This can be solved by passing in the textures filter setting to the resize_to_po2 code. We just need to add an optional boolean p_filter to resize_to_po2.

I might do this and send a PR request, I wanted to contribute to Godot for some time now.

**Comment by lawnjelly** [comment]
> I wonder if our POT codepath accidentally forces texture filtering.

This is possible. It sets the conditional `CanvasShaderGLES2::USE_FORCE_REPEAT` which may be causing a rebind and losing a filtering state. A graphics debugger would show this if we faked lack of non-pot support on desktop.

> What kind of workarounds? Are they really so bad? If so, we could make those workarounds optional, so that developers could decide for themselves if they want them.

Wrapping in the fragment shader. Yes, depending on the hardware. You can't really make them optional, unless you mean crash, display garbage, or display a black texture or something, if you are asking the hardware to do something it can't do. A GPU isn't like a general purpose computer (at least not GLES2), it has a limited set of functions.

It's kind of like asking a car to fly. Maybe you'd seen Elon Musk with a flying car, but you want to try it on your car. It can't fly, so it fakes it by displaying some sky on the windscreen as a workaround. You're not satisfied with the workaround, you turn it off and insist on it flying, you drive it off a cliff, and that won't go so well. :scream: 

What you can do is display a warning to the user to tell them they are doing something that could be problematic. This hasn't been implemented. So you try and put your car in fly mode, and it bleeps at you and puts up a message saying `WARNING! can't fly` etc. :smile:

**Comment by lawnjelly** [comment]
> We do resize the texture at load time

Ah that does explain it. But then there is no need to do the wrapping in the fragment shader. So the current arrangement is not good. Resizing the texture is preferable to the fragment shader approach imo.

**Comment by theogen-ratkin** [comment]
> We just need to add an optional boolean p_filter to resize_to_po2.

@clayjohn Wouldn't it be better to add a `p_interpolation` parameter like in `resize()` so that we could choose from all `Interpolation` members?

```
void resize_to_po2(bool p_square = false, Interpolation p_interpolation = INTERPOLATE_NEAREST);
void resize(int p_width, int p_height, Interpolation p_interpolation = INTERPOLATE_BILINEAR);
```

I tested `resize_to_po2` with `INTERPOLATE_NEAREST` and it's noticeably better, almost perfect, but you can see that it's stretched vertically a bit (since the image is resized from 512x104 to 512x128). Maybe this could be fixed by adding empty spaces to image, so that its aspect ratio is preserved?

![image](https://user-images.githubusercontent.com/20743993/102425009-d1cd8d00-3fe2-11eb-9cc5-cf242feeb9c3.png)

**Comment by clayjohn** [comment]
@theogen-ratkin 

> Wouldn't it be better to add a p_interpolation parameter like in resize() so that we could choose from all Interpolation members?

Yep. Good suggestion. 

Also, when calling from rasterizer_storage don't forget that it will always have to be nearest. I was wrong to say you should switch between linear and nearest. 

Regarding adding padding. I think stretching is a better result. Sure, it won't be good for pixel-perfect assets, but when using pixel perfect assets you need to be more careful designing them anyway. The majority of assets will be better off with a bit of stretching. And as you rightly point out, you would introduce space on wrapping and black bars.

**Comment by theogen-ratkin** [comment]
@clayjohn

Yes, good point about stretching, I guess it should be up to developers to make sure that textures are sized properly.

I have just sent pull request #44445 with my changes to `resize_to_po2()`.

**Comment by akien-mga** [comment]
Fixed by #44460.

**Comment by theogen-ratkin** [comment]
Awesome! Thanks for letting me make my first contribution to Godot!
